### PR TITLE
Implement visual and stability fixes

### PR DESCRIPTION
**Pull request by JvnSlv**
#### Overview

> Closes: #140 

This PR implements visual improvements

#### Checklist

*Implementation*
- [x] Implementation matches ticket acceptance criteria and technical notes
- [x] Manually tested against Acceptance Criteria
- [x] UI checked in Light / Dark mode

*Stability*
- [x] Checked if changes affect any features and verified affected features work as expected
- [ ] ~Added unit tests for new code and verified existing tests work as expected~

*Code quality*
- [ ] Updated `CHANGELOG.md`, `README.md` and package versions in `pubspec.yaml`
- [ ] ~Dependencies are updated to latest versions or new tickets are created if there are breaking changes or deprecations~
- [x] If an unrelated part of the codebase needs to be updated or refactored, create tickets with proposed changes

### CODE REVIEW

File: packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart [hunk 1]
```diff
@@ -60,6 +60,7 @@ class PinCodeBloc extends $PinCodeBloc {
   final bool autoPromptBiometric;
 
   final BehaviorSubject<String> _pinCode = BehaviorSubject.seeded('');
+  final BehaviorSubject<bool> isAuthenticated = BehaviorSubject.seeded(false);
```
Comment (StanevPrime) [inline packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart:63]: Do we really need to introduce a new subject for authentication tracking ? 

I believe here we have a fundamental issue. The `digitsCount` state should represent the `_pinCode` length, while `_$addDigitEvent` and `_$deleteDigitEvent` should control the `_pinCode` value.

**Review by StanevPrime — CHANGES_REQUESTED** [review]

File: packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart [hunk 2]
```diff
@@ -52,6 +52,19 @@ class PinCodeBloc extends $PinCodeBloc {
     required this.autoPromptBiometric,
   }) {
     authenticated.connect().addTo(_compositeSubscription);
+    _$deleteDigitEvent
+        .map((_) => _pinCode.add(_pinCode.value.substring(
```
Comment (StanevPrime) [inline packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart:56]: - The map should do only only thing ... map one value to another. No side effects should be executed here
- Please use one listener for updating the `_pinCode` subjects
- Please use functional programming where the operators do one thing (at the current implementation the switchMap do some checks and implicitly assigns new value to the `_pinCode`)

You can achieve explicit, declarative, functional implementation such as

```
Rx.merge([
    _$deleteDigitEvent.map(/* get the latest value from the  _pinCode and emit a new state of it */),
    _$addDigitEvent.switchMap(/* get the latest value from the  _pinCode and emit a new state of it */),
]).listen(_pinCode.add).dispose...
.
```

**Review by StanevPrime — CHANGES_REQUESTED** [review]

File: packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart [hunk 3]
```diff
@@ -52,6 +52,14 @@ class PinCodeBloc extends $PinCodeBloc {
     required this.autoPromptBiometric,
   }) {
     authenticated.connect().addTo(_compositeSubscription);
+    Rx.merge([
+      _$addDigitEvent.map((digit) {
+        _pinCode.add(_pinCode.value + digit);
```
Comment (StanevPrime) [inline packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart:57]: Do not violate the SRP

File: packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart [hunk 4]
```diff
@@ -52,6 +52,14 @@ class PinCodeBloc extends $PinCodeBloc {
     required this.autoPromptBiometric,
   }) {
     authenticated.connect().addTo(_compositeSubscription);
+    Rx.merge([
+      _$addDigitEvent.map((digit) {
```
Comment (StanevPrime) [inline packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart:56]: Do not remove the check for maximum pin length

File: packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart [hunk 5]
```diff
@@ -62,17 +70,11 @@ class PinCodeBloc extends $PinCodeBloc {
...
-          (_) {
-            _pinCode.add(_pinCode.value.substring(
-                0, _pinCode.value.isNotEmpty ? _pinCode.value.length - 1 : 0));
-            return Stream.value(_pinCode.value.length);
-          },
-        ).asResultStream(),
-        errorState.mapTo(0).asResultStream(),
-      ]).whereSuccess().startWith(0).share();
+  Stream<int> _mapToDigitsCountState() => _pinCode
+      .flatMap<int>((pinCode) => Stream.value(pinCode.length))
```
Comment (StanevPrime) [inline packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart:74]: Use flatMap only when dealing with streams, here the map works just fine

File: packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart [hunk 6]
```diff
@@ -89,7 +91,7 @@ class PinCodeBloc extends $PinCodeBloc {
 
   @override
   ConnectableStream<dynamic> _mapToAuthenticatedState() => Rx.merge([
-        _digitsCountState.switchMap((digitsCount) =>
+        digitsCount.flatMap((digitsCount) =>
```
Comment (StanevPrime) [inline packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart:94]: Use the where operator to ensure the authentication triggers at the right moment

File: packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart [hunk 7]
```diff
@@ -135,23 +137,17 @@ class PinCodeBloc extends $PinCodeBloc {
...
   Future<dynamic> _checkPin(String pinCode, int digits) async {
     final storedPinLength = await pinCodeService.getPinLength();
     if (storedPinLength != 0 && digits == storedPinLength) {
       try {
         final encryptedPin = await pinCodeService.encryptPinCode(pinCode);
         await pinCodeService.savePinCodeInSecureStorage(encryptedPin);
-        return await pinCodeService.verifyPinCode(encryptedPin);
+        final verificationResult =
+            await pinCodeService.verifyPinCode(encryptedPin);
+        _pinCode.value = '';
```
Comment (StanevPrime) [inline packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart:149]: Avoid assigning behaviour values in private functions. This should be explicitly implemented

**Review by StanevPrime — CHANGES_REQUESTED** [review]

File: packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart [hunk 8]
```diff
@@ -52,6 +52,11 @@ class PinCodeBloc extends $PinCodeBloc {
     required this.autoPromptBiometric,
   }) {
     authenticated.connect().addTo(_compositeSubscription);
+    Rx.merge([
```
Comment (StanevPrime) [inline packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart:55]: As mentioned the max length check should be applied on adding new digits

File: packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart [hunk 9]
```diff
@@ -62,17 +67,11 @@ class PinCodeBloc extends $PinCodeBloc {
...
-            _pinCode.add(_pinCode.value.substring(
-                0, _pinCode.value.isNotEmpty ? _pinCode.value.length - 1 : 0));
-            return Stream.value(_pinCode.value.length);
-          },
-        ).asResultStream(),
-        errorState.mapTo(0).asResultStream(),
-      ]).whereSuccess().startWith(0).share();
+  Stream<int> _mapToDigitsCountState() => _pinCode
+      .map<int>((pinCode) => pinCode.length)
+      .asResultStream()
```
Comment (StanevPrime) [inline packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart:72]: Usually the result stream is used on asynchronous calls. From what I see this is redundant

File: packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart [hunk 10]
```diff
@@ -89,8 +88,13 @@ class PinCodeBloc extends $PinCodeBloc {
 
   @override
   ConnectableStream<dynamic> _mapToAuthenticatedState() => Rx.merge([
-        _digitsCountState.switchMap((digitsCount) =>
-            _checkPin(_pinCode.value, digitsCount).asResultStream()),
+        digitsCount
+            .switchMap((_) => pinCodeService.getPinLength().asStream())
+            .where(
+                (storedPinLength) => storedPinLength == _pinCode.value.length)
+            .flatMap((digitsCount) => _checkPin(_pinCode.value)
```
Comment (StanevPrime) [inline packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart:95]: Why we use the flagMap instead of switch map?

**Review by StanevPrime — CHANGES_REQUESTED** [review]

File: packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart (hunk 10 above)
Comment (JvnSlv) [reply packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart]: No reason, reverted to switch map

File: packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart (hunk 9 above)
Comment (JvnSlv) [reply packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart]: Removed ResultStream

File: packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart (hunk 8 above)
Comment (JvnSlv) [reply packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart:55]: Added

File: packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart (hunk 7 above)
Comment (JvnSlv) [reply packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart]: Removed

File: packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart (hunk 6 above)
Comment (JvnSlv) [reply packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart]: Done

File: packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart (hunk 5 above)
Comment (JvnSlv) [reply packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart]: Done

File: packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart (hunk 4 above)
Comment (JvnSlv) [reply packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart]: Added

File: packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart (hunk 3 above)
Comment (JvnSlv) [reply packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart]: Changed

File: packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart (hunk 2 above)
Comment (JvnSlv) [reply packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart]: Refactored

File: packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart [hunk 11]
```diff
@@ -169,11 +153,21 @@ class PinCodeBloc extends $PinCodeBloc {
...
-        return await pinCodeService.verifyPinCode(pinCode);
+        final verificationResult = await pinCodeService.verifyPinCode(pinCode);
+        return verificationResult;
       }
     }
     return false;
   }
 
+  /// Adds a digit to the pin code value
+  Future<String> _addDigit(String digit) async {
```
Comment (StanevPrime) [inline packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart:164]: I think we can pick a method name that corresponds to the actual implementation. It seems that we do not add a digit to the pin code but rather compute a new digits based on some criteria. 

WIth that being said would suggest either renaming the method to reflect its implementation (suggestion: `_computeDigits`) or to use a chain of reactive operators in its usage such as

```
_$addDigitEvent
    .asyncMap((digit) async => (digit: digit:, allowedLength: await pinCodeService.getPinLength())))
    .where((data) => _pinCode.value.length <= data.allowedLength)
    .map((data) => _pinCode.value + data.digit)

```

**Review by StanevPrime — CHANGES_REQUESTED** [review]

File: packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart [hunk 12]
```diff
@@ -52,6 +52,11 @@ class PinCodeBloc extends $PinCodeBloc {
     required this.autoPromptBiometric,
   }) {
     authenticated.connect().addTo(_compositeSubscription);
+    Rx.merge([
+      _$addDigitEvent.switchMap((digit) => _addDigit(digit).asStream()),
+      _$deleteDigitEvent
+          .map((_) => _pinCode.value.substring(0, _pinCode.value.length - 1)),
+    ]).listen((event) => _pinCode.value = event).addTo(_compositeSubscription);
```
Comment (StanevPrime) [inline packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart:59]: Can we use method reference to improve the readability ?

**Review by StanevPrime — CHANGES_REQUESTED** [review]

File: packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart (hunk 11 above)
Comment (JvnSlv) [reply packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart]: Done

File: packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart (hunk 12 above)
Comment (JvnSlv) [reply packages/widget_toolkit_pin/lib/src/lib_pin_code_with_biometrics/blocs/pin_code_bloc.dart]: Yes, done

**Pull request merged**